### ISSUE TITLE
No echo in rescue shell

### ISSUE THREAD

**Issue by zhsj** [elves/elvish#232]
```
execing recovery shell /bin/sh
$ sh: turning off NDELAY mode

$ 
```

I can't input anything in recovery shell.

**Comment by huiyiqun** [comment]
You can, but no display for what has been typed.

**Comment by zhsj** [comment]
Oh, I can type `exit` to exit. But without seeing any word on the editor line is so confusing.

**Comment by xiaq** [comment]
This can be fixed by putting the terminal back in canonical mode when launching rescue shell.

**Comment by xiaq** [comment]
This happens because some panic happens outside the main goroutine, and the editor cleanup code in `defer` is not executed.

**Comment by krader1961** [comment]
@xiaq: It seems to me this has been resolved. I added some `println()` instrumentation as well as some explicit `panic()` calls and it appears to me the TTY mode restoration always occurs.

I also explored solutions such as running `stty sane` in the rescue shell. Sadly, that isn't possible to do in a reliable manner given that */bin/sh* might refer to any of several widely used shells (e.g., bash, dash, ksh etc.). If anything else needs to be done here it is probably just adding a `println("You might need to run: stty sane")` after the `println("\nExecuting rescue shell /bin/sh")`. It appears to me that elvish makes a reasonable attempt to restore the tty mode before launching the rescue shell.
